### PR TITLE
[INV-68]-Create method-to-create-accommodation

### PR DESCRIPTION
**Pull request by blackseriesC63**
Post Screenshot 2024-11-13 at 19 38 37
Result Screenshot 2024-11-13 at 19 38 44

### CODE REVIEW

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 1]
```diff
@@ -1,32 +1,61 @@
...
+  async createAccommodation(
+    @Body() createAccommodationDto: AccommodationDto,
+  ): Promise<Accommodation> {
+    return this.accommodationService.create(createAccommodationDto);
   }
 
-  @Get()
-  findAll() {
-    return this.AccommodationService.findAll();
+  // Route to create an address for an accommodation
```
Comment (NegatiV-64) [inline src/modules/accommodations/controllers/accommodations.controller.ts:31]: Please remove the comment

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 2]
```diff
@@ -1,32 +1,61 @@
...
+    @Param('accommodationId') accommodationId: string,
+    @Body() addressDto: AccommodationAddressDto,
+  ): Promise<AccommodationAddress> {
+    return this.accommodationAddressService.create(accommodationId, addressDto);
   }
 
-  @Get()
-  findOne() {
-    return this.AccommodationService.findOne();
+  // Route to add images to an accommodation
```
Comment (NegatiV-64) [inline src/modules/accommodations/controllers/accommodations.controller.ts:40]: Please remove the comment

File: src/modules/accommodations/services/accommodations.service.ts [hunk 3]
```diff
@@ -1,30 +1,66 @@
...
+          accommodation: savedAccommodation,
+        }),
+      );
+      await this.accommodationImageRepository.save(images);
+    }
 
-  update() {
-    return `This action updates Accommodation`;
-  }
+    if (createAccommodationDto.amenity) {
```
Comment (NegatiV-64) [inline src/modules/accommodations/services/accommodations.service.ts:56]: I think that this check is redundant

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 4]
```diff
@@ -1,32 +1,61 @@
...
+  ): Promise<Accommodation> {
+    return this.accommodationService.create(createAccommodationDto);
   }
 
-  @Get()
-  findAll() {
-    return this.AccommodationService.findAll();
+  // Route to create an address for an accommodation
+  @Post(':accommodationId/address')
+  async createAddress(
```
Comment (NegatiV-64) [inline src/modules/accommodations/controllers/accommodations.controller.ts:33]: Not sure that we need this endpoint. Address is created when accommodation is created

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 5]
```diff
@@ -1,32 +1,61 @@
...
-  update() {
-    return this.AccommodationService.update();
-  }
-
-  @Delete()
-  remove() {
-    return this.AccommodationService.remove();
+  // Route to add amenities to an accommodation
+  @Post(':accommodationId/amenity')
+  async createAmenity(
```
Comment (NegatiV-64) [inline src/modules/accommodations/controllers/accommodations.controller.ts:55]: Not sure that we need this endpoint. Amenities are created when accommodation is created

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+  @ApiProperty({ description: 'Date when the amenity was last updated' })
+  @IsDate()
+  @IsOptional()
+  updatedAt: Date;
+
+  @OneToOne(() => Accommodation, (accommodation) => accommodation.amenity, {
+    onDelete: 'CASCADE',
+  })
+  @JoinColumn({ name: 'accommodation_id' }) // Explicitly define the column name here
```
Comment (NegatiV-64) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:113]: Remove the comments

Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:113]: Please remove comments

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+  accommodation: Accommodation;
+
+  @Column({ name: 'accommodation_id' }) // Explicitly define the accommodation_id column
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:116]: Please remove comments

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
+import { ApiProperty } from '@nestjs/swagger';
+import { IsBoolean, IsDate, IsOptional } from 'class-validator';
+import { Column, JoinColumn, OneToOne } from 'typeorm';
+import { Accommodation } from '#/modules/accommodations/entities/accommodations.entity';
+
+export class AccommodationAmenityDto {
+  @ApiProperty({ description: 'Indicates if the accommodation has Wi-Fi' })
+  @IsBoolean()
+  hasWifi: boolean;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:9]: Please make this optional too

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+  @ApiProperty({ description: 'Indicates if the accommodation has parking' })
+  @IsBoolean()
+  hasParking: boolean;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:13]: Please make this optional too

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    description: 'Indicates if the accommodation is in a quiet area',
+  })
+  @IsBoolean()
+  @IsOptional()
+  isQuietArea: boolean;
+
+  @ApiProperty({ description: 'Date when the amenity was created' })
+  @IsDate()
+  @IsOptional()
+  createdAt: Date;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:103]: Please remove this field

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+  @ApiProperty({ description: 'Date when the amenity was last updated' })
+  @IsDate()
+  @IsOptional()
+  updatedAt: Date;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:108]: Please remove this field

File: src/modules/accommodations/services/accommodations.service.ts [hunk 12]
```diff
@@ -1,30 +1,66 @@
...
+    const accommodation = this.accommodationRepository.create(
+      createAccommodationDto,
+    );
+    const savedAccommodation =
+      await this.accommodationRepository.save(accommodation);
 
-  findAll() {
-    return `This action returns all Accommodation`;
-  }
+    if (createAccommodationDto.address) {
```
Comment (NegatiV-64) [inline src/modules/accommodations/services/accommodations.service.ts:35]: This check is redundant

File: src/modules/accommodations/services/accommodations.service.ts [hunk 13]
```diff
@@ -1,30 +1,66 @@
...
-  create() {
-    return 'This action adds a new Accommodation';
-  }
+  async create(
+    createAccommodationDto: AccommodationDto,
+  ): Promise<Accommodation> {
+    const accommodation = this.accommodationRepository.create(
+      createAccommodationDto,
+    );
+    const savedAccommodation =
```
Comment (NegatiV-64) [inline src/modules/accommodations/services/accommodations.service.ts:32]: Also we need to save the amenities right here

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]
Good job, left some comments + add Swagger API support for the responses

File: src/modules/accommodations/dto/requests/create-accommodation.req.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  @ApiProperty({ description: 'Name of the accommodation' })
+  @IsString()
+  name: string;
+
+  @ApiProperty({ description: 'Description of the accommodation' })
+  @IsString()
+  description: string;
+
+  @ApiProperty({ description: 'Cover image URL of the accommodation' })
+  @IsOptional() // Make this optional if it might not always be provided
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation.req.ts:27]: Please remove comments

File: src/modules/accommodations/dto/requests/create-accommodation.req.ts [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+  @Min(1)
+  numberOfRooms: number;
+
+  @ApiProperty({ description: 'Allowed number of people in the accommodation' })
+  @IsNumber()
+  allowedNumberOfPeople: number;
+
+  @ApiProperty({ description: "Accommodation owner's user ID" })
+  @IsUUID()
+  ownerId: string;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation.req.ts:65]: I guess we will take owner ID from request.user

File: src/modules/accommodations/entities/accommodation-address.entity.ts [hunk 16]
```diff
@@ -14,22 +14,22 @@ export class AccommodationAddress {
...
   city: string;
 
-  @Column()
+  @Column({ type: 'varchar' })
   country: string;
 
-  @Column({ nullable: true, name: 'zip_code' })
-  zipCode: string;
+  @Column({ type: 'varchar' })
+  zip_code: string;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/entities/accommodation-address.entity.ts:27]: Please change it back. Old version was correct

**Review by ShahobiddinIsmoilov — CHANGES_REQUESTED** [review]

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 17]
```diff
@@ -1,32 +1,42 @@
...
+    description: 'Accommodation created successfully.',
+    type: Accommodation,
+  })
+  @ApiResponse({ status: 400, description: 'Invalid data provided.' })
+  @ApiResponse({ status: 401, description: 'Unauthorized request.' })
+  async create(
+    @Body() accommodationDto: AccommodationDto,
+    @GetAccount('accountId') ownerId: string,
+  ): Promise<Accommodation> {
+    if (!ownerId) {
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/controllers/accommodations.controller.ts:34]: I think this check is unnecessary

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 18]
```diff
@@ -1,32 +1,42 @@
...
+      throw new Error('Account ID is missing in the request');
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/controllers/accommodations.controller.ts:35]: Also, wrong handling of the error. You should throw ```UnauthorizedException```, not pure ```Error```

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 19]
```diff
@@ -1,32 +1,42 @@
...
+  ApiTags,
+  ApiOperation,
+  ApiBearerAuth,
+  ApiResponse,
+} from '@nestjs/swagger';
+import { AccommodationService } from '#/modules/accommodations/services/accommodations.service';
+import { Accommodation } from '#/modules/accommodations/entities/accommodations.entity';
+import { AccommodationDto } from '#/modules/accommodations/dto/requests/create-accommodation.req';
+import { AccessTokenGuard } from '#/shared/guards/access-token.guard';
+import { GetAccount } from '../../auth/decorators/get-account.decorator';
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/controllers/accommodations.controller.ts:12]: Bad import path

File: src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+    onDelete: 'CASCADE',
+  })
+  @JoinColumn({ name: 'accommodation_id' })
+  accommodation: Accommodation;
+
+  @Column({ name: 'accommodation_id' })
+  accommodationId: string;
+}
+
+AccommodationAmenityDto;
```
Comment (ShahobiddinIsmoilov) [inline src/modules/accommodations/dto/requests/create-accommodation-amenity.req.ts:112]: Why do we need this?

**Review by ShahobiddinIsmoilov — CHANGES_REQUESTED** [review]

**Review by ShahobiddinIsmoilov — APPROVED** [review]

**Review by ShahobiddinIsmoilov — DISMISSED** [review]

File: src/modules/accommodations/controllers/accommodations.controller.ts [hunk 21]
```diff
@@ -1,32 +1,39 @@
...
+import { AccommodationDto } from '#/modules/accommodations/dto/requests/create-accommodation.req';
+import { AccessTokenGuard } from '#/shared/guards/access-token.guard';
+import { GetAccount } from '#/modules/auth/decorators/get-account.decorator';
 
-@Controller('accommodation')
-export class AccommodationsController {
-  constructor(private readonly AccommodationService: AccommodationsService) {}
+@ApiTags('Accommodations')
+@ApiBearerAuth()
+@UseGuards(AccessTokenGuard)
```
Comment (NegatiV-64) [inline src/modules/accommodations/controllers/accommodations.controller.ts:16]: Move guard to the methods level. We don't need to guard every route here

Comment (NegatiV-64) [inline src/modules/accommodations/controllers/accommodations.controller.ts:16]: Move guard to the methods level. We don't need to guard every route here

File: src/modules/accommodations/services/accommodations.service.ts [hunk 22]
```diff
@@ -1,30 +1,53 @@
...
-    return `This action updates Accommodation`;
-  }
+    await this.accommodationImageService.create(
+      createdAccommodation.id,
+      createAccommodationDto.images,
+    );
 
-  remove() {
-    return `This action removes Accommodation`;
+    return createdAccommodation;
```
Comment (NegatiV-64) [inline src/modules/accommodations/services/accommodations.service.ts:51]: Would it have the address, images & amenities?

**Review by NegatiV-64 — CHANGES_REQUESTED** [review]

**Review by NegatiV-64 — APPROVED** [review]

**Pull request merged**